helix-core/src/history.rs

```diff
@@ -630,4 +875,151 @@ mod test {
             Err("duration too large".to_string())
         );
     }
+
+    fn is_tree(h: &History) -> bool {
+        use ahash::{AHashMap, AHashSet};
+
+        let n = h.revisions.len();
+        let nodes = &h.revisions;
+
+        if n == 0 {
+            return false;
+        }
+
+        let mut adj_list = AHashMap::with_capacity(n);
+        for (node, parent) in nodes.iter().enumerate().map(|(idx, r)| (idx, r.parent)) {
+            // Skip loop
+            if !(node == 0 && parent == 0) {
+                adj_list
+                    .entry(node)
+                    .or_insert_with(AHashSet::new)
+                    .insert(parent);
+                adj_list
+                    .entry(parent)
+                    .or_insert_with(AHashSet::new)
+                    .insert(node);
+            }
+        }
+
+        let mut visited = AHashSet::new();
+        let mut stack = vec![0];
+        while let Some(node) = stack.pop() {
+            if !visited.insert(node) {
+                continue;
+            }
+
+            if let Some(adj_nodes) = adj_list.get(&node) {
+                for v in adj_nodes {
+                    if !visited.contains(v) {
+                        stack.push(*v);
+                    }
+                }
+            }
+        }
+        visited.len() == n
+    }
+
+    fn generate_history(mut inserts: Vec<String>) -> (History, Rope) {
+        use rand::distributions::{Distribution, Uniform};
+        use rand::SeedableRng;
+
+        let dist = Uniform::new_inclusive(0, 2);
+        let mut rng = rand::rngs::SmallRng::from_entropy();
+
+        let mut hist = History::default();
+        let mut doc = Rope::default();
+        let sel = Selection::point(0);
+
+        while !inserts.is_empty() {
```

Comment:
Self-note: This is a bad test implementation because it commits a new revision more often than it performs some other action.